# ------------------------------------------------------------------------------
# This code is part of paper: In silico Mapping of Non-Canonical DNA Structures Across the Human Ribosomal DNA Locus.
# Author: Jyoti Devendra Adala under supervision of Dr. Bruce Knutson
# For updates and contributions, visit : https://github.com/adalaj
#
# Purpose:
# This R script generates genome-wide visualizations of R-loop forming sequences (RLFS) 
# predicted within the human ribosomal DNA (rDNA) locus.
# It uses the karyoploteR package to produce ideogram-style plots showing the spatial distribution of RLFS along rDNA subregions.
#
# Specifically, it:
#  1. Reads input RLFS predictions generated by QmRLFS finder algorithm
#  2. Separates template and non-template strands for independent visualization.
#  3. Defines rDNA subregions (Promoter, 5′ETS, 18S, ITS1, 5.8S, ITS2, 28S, 3′ETS, and IGS) 
#with custom coordinates and color coding.
#  4. Plots RLFS positions and coverage for each strand across the rDNA locus, highlighting structural organization and density patterns.
#  5. Exports publication-quality figures (TIFF/PNG) illustrating strandwise and full-locus distributions.
#
# Inputs:
#   - Human rDNA FASTA sequence (GenBank: KY962518)
#   - RLFS BED output file from QmRLFS-finder (v1.5)
#
# Outputs:
#   - High-resolution ideogram plots (.png, .tiff) showing RLFS locations across the human rDNA locus
#   - Separate visualizations for template and non-template strands
#
# ------------------------------------------------------------------------------


#load libraries
library(data.table)
library(tidyverse)
library(BiocManager)
BiocManager::install("karyoploteR")
library(karyoploteR)


##Two ways to do it:
#1) using karyoploteR

#read the rlfs that overlapped with rdna locus
rdna<- fread("KY962518_1_humanrDNA_2018_qmrlfs.out.bed", sep = "\t", header = FALSE)
rdna$V1= "rDNA_locus"
rdna6<- rdna %>% select(1:6)
colnames(rdna6)<- c("chr", "start", "end", "name", "score", "strand")

##separate as per strand
rdna6_nontemplate<- rdna6 %>% filter(strand=="+")
#because in NCBI keep nontemplate sequence.
rdna6_template<- rdna6 %>% filter(strand=="-")



#Plus Strand R-loops (Non-template Strand):
#R-loops annotated as being on the plus strand (non-template strand) indicate RNA is hybridized to the minus strand (template strand).
#These R-loops are formed during sense transcription, where the RNA transcript is complementary to the DNA template strand.

#Minus Strand R-loops (Template Strand):
#R-loops annotated as being on the minus strand (template strand) indicate RNA is hybridized to the plus strand (non-template strand).
#These R-loops are formed during antisense transcription, where RNA transcripts are complementary to the non-template strand.




 ##plotting with IGS on start and promoter at end.
 # i manually added 3500bp from IGS at the front. Out of which, first 1298 belongs to IGS and remaining 2202bp will be considered as promoter. 
 # rationale is that I took the sequences form hg38 and checked how many nucleotides match with IGS. I realized only 2200 match from hg38 to IGS. 
 # so the total length of KY962518_added_3500nt_IGS_upstream_nontemplate.fasta is 44838+3500 = 48338
 
 ###Plot4 showing entire rDNA and begining of next
 
 #open terminal
 #(base)  Downloads % conda activate python2.7
 #(python2.7) Downloads % python QmRLFS-finder.py -bed -i KY962518_added_3500nt_IGS_upstream_nontemplate.fasta -o KY962518_added_3500nt_IGS_upstream_qmrlfs
 #QmRLFS-finder.py (version v1.5)
 #run on Fri Feb 21 2025 05:16:52 
 #command line: python QmRLFS-finder.py -bed -i KY962518_added_3500nt_IGS_upstream_nontemplate.fasta -o KY962518_added_3500nt_IGS_upstream_qmrlfs
 
 
 #read the rlfs that overlapped with rdna locus
 entire_rdna<- fread("KY962518_added_3500nt_IGS_upstream_qmrlfs.out.bed", sep = "\t", header = FALSE) #208, this contain double entry for promoter and IGS. 
  
 
 entire_rdna$V1= "rDNA_locus"
 entire_rdna6<- entire_rdna %>% select(1:6)
 colnames(entire_rdna6)<- c("chr", "start", "end", "name", "score", "strand")
 
 ##separate as per strand
 entire_rdna6_nontemplate<- entire_rdna6 %>% filter(strand=="+") #94
 #because in NCBI keep nontemplate sequence.
 
 
 entire_rdna6_template<- entire_rdna6 %>% filter(strand=="-")#114
 

 png("rdna_both_strand_rlfs.png", width = 15, height= 10, units= "in", res = 600) #png is supported by biorender
 
 ##plotting begins
 custom_genome <- toGRanges(data.frame(chr="rDNA_locus", start=1, end=48338))
 #end is 48338 because I added 3500 to 44838
 #changing  colors 

 kp <- plotKaryotype(genome=custom_genome, plot.type = 2)
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 1, x1 =1298 , y0 = 0, y1 = 1, col = "#A4A2A8", data.panel = "ideogram", borders= NA) #marks last 1298bp from IGS representing previous rdna 
 kpRect(kp, chr = 'rDNA_locus', x0 = 1299, x1 =3500 , y0 = 0, y1 = 1, col = "#B6FFF4", data.panel = "ideogram", borders= NA) #marks 2202 bp of  promoter
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 3501, x1 = 7157 , y0 = 0, y1 = 1, col = "#FDCCE5", data.panel = "ideogram", borders= NA) #marks 5'ETS (3501+(3657-1))
 #3501+(3657-1) = 7157
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 7158, x1 = 9026, y0 = 0, y1 = 1, col = "#D0B6FF", data.panel = "ideogram", borders= NA) #marks 18S
 #7158+ (1869-1) = 9026
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 9027, x1 = 10096, y0 = 0, y1 = 1, col = "#EF9B20", data.panel = "ideogram", borders= NA) #marks ITS1S
 #9027+ (1070-1) = 10096 
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 10097, x1 = 10253, y0 = 0, y1 = 1, col = "#A0322B", data.panel = "ideogram", borders= NA) #marks 5.8S
 #10097+ (157-1) = 10253
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 10254, x1 = 11420, y0 = 0, y1 = 1, col = "#FFCC17", data.panel = "ideogram", borders= NA) #marks ITS2
 #10254+(1167-1) = 11420
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 11421, x1 = 16471, y0 = 0, y1 = 1, col = "#E5FFB6", data.panel = "ideogram", borders= NA) #marks 28S
 #11421+(5051-1) = 16471
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 16472, x1 = 16832, y0 = 0, y1 = 1, col = "#3B8CC4", data.panel = "ideogram", borders= NA) #marks 3'ETS
 #16472+(361-1) = 16832
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 16833, x1 = 48338, y0 = 0, y1 = 1, col = "#A4A2A8", data.panel = "ideogram", borders= NA) #marks IGS
 #16833+ (31506-1)= 48338
 
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 46137, x1 = 48338, y0 = 0, y1 = 1, col = "#B6FFF4", data.panel = "ideogram", borders= NA) #marks 2202bp of promoter
 #48338-2201
 
 kpPlotRegions(kp, data=entire_rdna6_template, col="#1414E1", r0= -0.5, r1= -1.9)
 kpPlotRegions(kp, data=entire_rdna6_nontemplate, col="#E21515", r0= -0.5, r1= -1.3) #-1.5 to make blue with more width
 dev.off()

 #use zoom option, took screenshot and edited in powerpoint
 
 
 #for checking how many RLFS are formed after rule count please see junction_rloop_2018.R file
 
 
 
 

 #need to plot only from 5'ETS to 3'ETS
 png("rdna_nontemplate_rlfs_coverage.png", width = 10, height= 10, units= "in", res = 600)
 
 custom_genome <- toGRanges(data.frame(chr="rDNA_locus", start=1, end=19000))
 kp <- plotKaryotype(genome=custom_genome, plot.type = 2)
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 1, x1 =1298 , y0 = 0, y1 = 1, col = "#A4A2A8", data.panel = "ideogram", borders= NA) #marks last 1298bp from IGS representing previous rdna 
 kpRect(kp, chr = 'rDNA_locus', x0 = 1299, x1 =3500 , y0 = 0, y1 = 1, col = "#B6FFF4", data.panel = "ideogram", borders= NA) #marks 2202 bp of  promoter
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 3501, x1 = 7157 , y0 = 0, y1 = 1, col = "#FDCCE5", data.panel = "ideogram", borders= NA) #marks 5'ETS (3501+(3657-1))
 #3501+(3657-1) = 7157
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 7158, x1 = 9026, y0 = 0, y1 = 1, col = "#D0B6FF", data.panel = "ideogram", borders= NA) #marks 18S
 #7158+ (1869-1) = 9026
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 9027, x1 = 10096, y0 = 0, y1 = 1, col = "#EF9B20", data.panel = "ideogram", borders= NA) #marks ITS1S
 #9027+ (1070-1) = 10096 
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 10097, x1 = 10253, y0 = 0, y1 = 1, col = "#A0322B", data.panel = "ideogram", borders= NA) #marks 5.8S
 #10097+ (157-1) = 10253
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 10254, x1 = 11420, y0 = 0, y1 = 1, col = "#FFCC17", data.panel = "ideogram", borders= NA) #marks ITS2
 #10254+(1167-1) = 11420
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 11421, x1 = 16471, y0 = 0, y1 = 1, col = "#E5FFB6", data.panel = "ideogram", borders= NA) #marks 28S
 #11421+(5051-1) = 16471
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 16472, x1 = 16832, y0 = 0, y1 = 1, col = "#3B8CC4", data.panel = "ideogram", borders= NA) #marks 3'ETS
 #16472+(361-1) = 16832
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 16833, x1 = 19000, y0 = 0, y1 = 1, col = "#A4A2A8", data.panel = "ideogram", borders= NA) #marks IGS
 
 #16472+(361-1) = 16832
 
 kpPlotCoverage(kp, data=entire_rdna6_nontemplate, col = "#E21515", lwd =5)
 kpPlotRegions(kp, data=entire_rdna6_nontemplate, data.panel=2, col = "#E21515")
 dev.off()
 
 
 #need to plot only from 5'ETS to 3'ETS
 png("rdna_template_rlfs_coverage.png", width = 10, height= 10, units= "in", res = 600)
 
 custom_genome <- toGRanges(data.frame(chr="rDNA_locus", start=1, end=19000))
 kp <- plotKaryotype(genome=custom_genome, plot.type = 2)
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 1, x1 =1298 , y0 = 0, y1 = 1, col = "#A4A2A8", data.panel = "ideogram", borders= NA) #marks last 1298bp from IGS representing previous rdna 
 kpRect(kp, chr = 'rDNA_locus', x0 = 1299, x1 =3500 , y0 = 0, y1 = 1, col = "#B6FFF4", data.panel = "ideogram", borders= NA) #marks 2202 bp of  promoter
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 3501, x1 = 7157 , y0 = 0, y1 = 1, col = "#FDCCE5", data.panel = "ideogram", borders= NA) #marks 5'ETS (3501+(3657-1))
 #3501+(3657-1) = 7157
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 7158, x1 = 9026, y0 = 0, y1 = 1, col = "#D0B6FF", data.panel = "ideogram", borders= NA) #marks 18S
 #7158+ (1869-1) = 9026
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 9027, x1 = 10096, y0 = 0, y1 = 1, col = "#EF9B20", data.panel = "ideogram", borders= NA) #marks ITS1
 #9027+ (1070-1) = 10096 
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 10097, x1 = 10253, y0 = 0, y1 = 1, col = "#A0322B", data.panel = "ideogram", borders= NA) #marks 5.8S
 #10097+ (157-1) = 10253
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 10254, x1 = 11420, y0 = 0, y1 = 1, col = "#FFCC17", data.panel = "ideogram", borders= NA) #marks ITS2
 #10254+(1167-1) = 11420
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 11421, x1 = 16471, y0 = 0, y1 = 1, col = "#E5FFB6", data.panel = "ideogram", borders= NA) #marks 28S
 #11421+(5051-1) = 16471
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 16472, x1 = 16832, y0 = 0, y1 = 1, col = "#3B8CC4", data.panel = "ideogram", borders= NA) #marks 3'ETS
 #16472+(361-1) = 16832
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 16833, x1 = 19000, y0 = 0, y1 = 1, col = "#A4A2A8", data.panel = "ideogram", borders= NA) #marks IGS
 #16472+(361-1) = 16832
 
 kpPlotCoverage(kp, data=entire_rdna6_template, col = "#1414E1", lwd =5)
 kpPlotRegions(kp, data=entire_rdna6_template, data.panel=2, col = "#1414E1")
 dev.off()
 
 #visually i dot think one can differentiate where ITS2 RLFS is beginning, it seems we have alot of ITS1 RLFS.
 #to see that 
 
 #i made three seperate garphs
 its2_test <- entire_rdna6_nontemplate %>%
               filter(start > 10254 & start < 11421)
 
 its1_test<- entire_rdna6_nontemplate %>%
              filter(start > 9027 & start < 10097)
 
 its1and2_test<- entire_rdna6_nontemplate %>%
                  filter(start > 9027 & start < 11421)
 
 custom_genome <- toGRanges(data.frame(chr="rDNA_locus", start=1, end=19000))
 kp <- plotKaryotype(genome=custom_genome, plot.type = 2)
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 1, x1 =1298 , y0 = 0, y1 = 1, col = "#A4A2A8", data.panel = "ideogram", borders= NA) #marks last 1298bp from IGS representing previous rdna 
 kpRect(kp, chr = 'rDNA_locus', x0 = 1299, x1 =3500 , y0 = 0, y1 = 1, col = "#B6FFF4", data.panel = "ideogram", borders= NA) #marks 2202 bp of  promoter
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 3501, x1 = 7157 , y0 = 0, y1 = 1, col = "#FDCCE5", data.panel = "ideogram", borders= NA) #marks 5'ETS (3501+(3657-1))
 #3501+(3657-1) = 7157
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 7158, x1 = 9026, y0 = 0, y1 = 1, col = "#D0B6FF", data.panel = "ideogram", borders= NA) #marks 18S
 #7158+ (1869-1) = 9026
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 9027, x1 = 10096, y0 = 0, y1 = 1, col = "#EF9B20", data.panel = "ideogram", borders= NA) #marks ITS1S
 #9027+ (1070-1) = 10096 
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 10097, x1 = 10253, y0 = 0, y1 = 1, col = "#A0322B", data.panel = "ideogram", borders= NA) #marks 5.8S
 #10097+ (157-1) = 10253
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 10254, x1 = 11420, y0 = 0, y1 = 1, col = "#FFCC17", data.panel = "ideogram", borders= NA) #marks ITS2
 #10254+(1167-1) = 11420
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 11421, x1 = 16471, y0 = 0, y1 = 1, col = "#E5FFB6", data.panel = "ideogram", borders= NA) #marks 28S
 #11421+(5051-1) = 16471
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 16472, x1 = 16832, y0 = 0, y1 = 1, col = "#3B8CC4", data.panel = "ideogram", borders= NA) #marks 3'ETS
 #16472+(361-1) = 16832
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 16833, x1 = 19000, y0 = 0, y1 = 1, col = "#A4A2A8", data.panel = "ideogram", borders= NA) #marks IGS
 
 #16472+(361-1) = 16832
 
 
 kpPlotCoverage(kp, data=its1_test, col = "#E21515")
 kpPlotRegions(kp, data=its1_test, data.panel=2, col = "#E21515")
 
 kpPlotCoverage(kp, data=its1and2_test, col = "#E21515")
 kpPlotRegions(kp, data=its1and2_test, data.panel=2, col = "#E21515")
 
 kpPlotCoverage(kp, data=its2_test, col = "#E21515")
 kpPlotRegions(kp, data=its2_test, data.panel=2, col = "#E21515")
 
 # i just took teh screenshot not saved as pdf. 
 
 #after looking at it i can see that in the final image, you can differentiate the the one that starts from ITS2 are actually present at the bottom
 #so last 7 RLFS between ITS1 to IT2 actually belong to ITS2
 #bruce told table will have cooridnates so we dont need to show separate ITS RLFS
 
 
 
 
 #I am trying to make bedgraphs directly in karyoplote
 ##plotting begins for entire region till IGS
 
 entire_rdna<- fread("KY962518_added_3500nt_IGS_upstream_qmrlfs.out.bed", sep = "\t", header = FALSE) #208, this contain double entry for promoter and IGS. 
 entire_rdna$V1= "rDNA_locus"
 entire_rdna6<- entire_rdna %>% select(1:6)
 colnames(entire_rdna6)<- c("chr", "start", "end", "name", "score", "strand")
 
 ##separate as per strand
 entire_rdna6_nontemplate<- entire_rdna6 %>% filter(strand=="+") #94
 #because in NCBI keep nontemplate sequence.
 
 
 entire_rdna6_template<- entire_rdna6 %>% filter(strand=="-")#114
 
 
 ##non template
 png("entire_rdna_nontemplate_rlfs_coverage.png", width = 30, height= 30, units= "in", res = 150)
 custom_genome <- toGRanges(data.frame(chr="rDNA_locus", start=1, end=48338))
 kp <- plotKaryotype(genome=custom_genome, plot.type = 2)
 
kpRect(kp, chr = 'rDNA_locus', x0 = 1, x1 =1298 , y0 = 0, y1 = 1, col = "#A4A2A8", data.panel = "ideogram", borders= NA) #marks last 1298bp from IGS representing previous rdna 
 kpRect(kp, chr = 'rDNA_locus', x0 = 1299, x1 =3500 , y0 = 0, y1 = 1, col = "#B6FFF4", data.panel = "ideogram", borders= NA) #marks 2202 bp of  promoter
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 3501, x1 = 7157 , y0 = 0, y1 = 1, col = "#FDCCE5", data.panel = "ideogram", borders= NA) #marks 5'ETS (3501+(3657-1))
 #3501+(3657-1) = 7157
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 7158, x1 = 9026, y0 = 0, y1 = 1, col = "#D0B6FF", data.panel = "ideogram", borders= NA) #marks 18S
 #7158+ (1869-1) = 9026
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 9027, x1 = 10096, y0 = 0, y1 = 1, col = "#EF9B20", data.panel = "ideogram", borders= NA) #marks ITS1S
 #9027+ (1070-1) = 10096 
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 10097, x1 = 10253, y0 = 0, y1 = 1, col = "#A0322B", data.panel = "ideogram", borders= NA) #marks 5.8S
 #10097+ (157-1) = 10253
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 10254, x1 = 11420, y0 = 0, y1 = 1, col = "#FFCC17", data.panel = "ideogram", borders= NA) #marks ITS2
 #10254+(1167-1) = 11420
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 11421, x1 = 16471, y0 = 0, y1 = 1, col = "#E5FFB6", data.panel = "ideogram", borders= NA) #marks 28S
 #11421+(5051-1) = 16471
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 16472, x1 = 16832, y0 = 0, y1 = 1, col = "#3B8CC4", data.panel = "ideogram", borders= NA) #marks 3'ETS
 #16472+(361-1) = 16832
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 16833, x1 = 48338, y0 = 0, y1 = 1, col = "#A4A2A8", data.panel = "ideogram", borders= NA) #marks IGS
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 46137, x1 = 48338, y0 = 0, y1 = 1, col = "#B6FFF4", data.panel = "ideogram", borders= NA) #marks 2202bp of promoter
 #48338-2201
 #16472+(361-1) = 16832
 kpPlotCoverage(kp, data=entire_rdna6_nontemplate, col = "#E21515")
 kpPlotRegions(kp, data=entire_rdna6_nontemplate, data.panel=2, col = "#E21515")
 dev.off()
 
 
 #template
 png("entire_rdna_template_rlfs_coverage.png", width = 30, height= 30, units= "in", res = 150)
 custom_genome <- toGRanges(data.frame(chr="rDNA_locus", start=1, end=48338))
 kp <- plotKaryotype(genome=custom_genome, plot.type = 2)
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 1, x1 =1298 , y0 = 0, y1 = 1, col = "#A4A2A8", data.panel = "ideogram", borders= NA) #marks last 1298bp from IGS representing previous rdna 
 kpRect(kp, chr = 'rDNA_locus', x0 = 1299, x1 =3500 , y0 = 0, y1 = 1, col = "#B6FFF4", data.panel = "ideogram", borders= NA) #marks 2202 bp of  promoter
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 3501, x1 = 7157 , y0 = 0, y1 = 1, col = "#FDCCE5", data.panel = "ideogram", borders= NA) #marks 5'ETS (3501+(3657-1))
 #3501+(3657-1) = 7157
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 7158, x1 = 9026, y0 = 0, y1 = 1, col = "#D0B6FF", data.panel = "ideogram", borders= NA) #marks 18S
 #7158+ (1869-1) = 9026
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 9027, x1 = 10096, y0 = 0, y1 = 1, col = "#EF9B20", data.panel = "ideogram", borders= NA) #marks ITS1S
 #9027+ (1070-1) = 10096 
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 10097, x1 = 10253, y0 = 0, y1 = 1, col = "#A0322B", data.panel = "ideogram", borders= NA) #marks 5.8S
 #10097+ (157-1) = 10253
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 10254, x1 = 11420, y0 = 0, y1 = 1, col = "#FFCC17", data.panel = "ideogram", borders= NA) #marks ITS2
 #10254+(1167-1) = 11420
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 11421, x1 = 16471, y0 = 0, y1 = 1, col = "#E5FFB6", data.panel = "ideogram", borders= NA) #marks 28S
 #11421+(5051-1) = 16471
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 16472, x1 = 16832, y0 = 0, y1 = 1, col = "#3B8CC4", data.panel = "ideogram", borders= NA) #marks 3'ETS
 #16472+(361-1) = 16832
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 16833, x1 = 48338, y0 = 0, y1 = 1, col = "#A4A2A8", data.panel = "ideogram", borders= NA) #marks IGS
 
 kpRect(kp, chr = 'rDNA_locus', x0 = 46137, x1 = 48338, y0 = 0, y1 = 1, col = "#B6FFF4", data.panel = "ideogram", borders= NA) #marks 2202bp of promoter
 #48338-2201
 #16472+(361-1) = 16832
 
 kpPlotCoverage(kp, data=entire_rdna6_template, col = "#1414E1")
 kpPlotRegions(kp, data=entire_rdna6_template, data.panel=2, col = "#1414E1")
 dev.off()
 
 

 
 
 
 
 
 
 
 
 
 
 